### PR TITLE
Add animated particle sprites

### PR DESCRIPTION
**Pull request by thecodewarrior**

### CODE REVIEW

File: src/example/java/com/teamwizardry/librarianlib/test/particlesystem/examples/AnimatedParticles.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import net.minecraft.util.ResourceLocation
+import net.minecraft.util.math.Vec3d
+import org.apache.commons.lang3.RandomUtils
+import java.awt.Color
+
+object AnimatedParticles : ParticleExample {
+    var counter = 0
+    override fun update(origin: Vec3d) {
+//        AnimatedSystem.reload()
+//        for(i in 0 until 5) {
```
Comment (Bluexin) [inline src/example/java/com/teamwizardry/librarianlib/test/particlesystem/examples/AnimatedParticles.kt]:
```suggestion
```
Comments left in

File: src/example/java/com/teamwizardry/librarianlib/test/particlesystem/examples/AnimatedParticles.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            spriteIndex = EaseBinding(lifetime, age, easing = Easing.linear, bindingSize = 1, origin = startIndex, target = endIndex)
+        ))
+    }
+
+    fun spawn(lifetime: Double, size: Double, pos: Vec3d, color: Color, startIndex: Int, endIndex: Int) {
+        this.addParticle(lifetime,
+            size,
+            pos.x, pos.y, pos.z,
+            pos.x, pos.y, pos.z,
+//            RandomUtils.nextDouble(0.0, 0.6)-0.3, RandomUtils.nextDouble(0.25, 0.5), RandomUtils.nextDouble(0.0, 0.6)-0.3, // velocity
```
Comment (Bluexin) [inline src/example/java/com/teamwizardry/librarianlib/test/particlesystem/examples/AnimatedParticles.kt]:
```suggestion
```
Comment left in

File: src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/SpriteRenderModule.kt [hunk 3]
```diff
@@ -64,7 +64,15 @@ class SpriteRenderModule @JvmOverloads constructor(
...
-        @JvmField val enableBlend: Boolean = true
+        @JvmField val enableBlend: Boolean = true,
+        /**
+         * The size of the sprite sheet (must be a power of 2)
+         */
+        @JvmField val spriteSheetSize: Int = 1,
+        /**
+         * The sprite index (indexed left-to-right, top-to-bottom)
+         */
+        @JvmField val spriteIndex: ReadParticleBinding = ConstantBinding(0.0)
```
Comment (Bluexin) [inline src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/SpriteRenderModule.kt:75]: Is this binary compatible?
It might be thanks to `@JvmOverloads` but please check it really is

File: src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/VelocityUpdateModule.kt [hunk 4]
```diff
@@ -32,6 +32,7 @@ class VelocityUpdateModule(
         position.load(particle)
         if(previousPosition != null) {
             position.contents.copyInto(previousPosition.contents)
+            previousPosition.store(particle)
```
Comment (Bluexin) [inline src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/VelocityUpdateModule.kt:35]: This seems unrelated to current PR, it's a fix I assume?

**Review by Bluexin — CHANGES_REQUESTED** [review]

File: src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/VelocityUpdateModule.kt (hunk 4 above)
Comment (thecodewarrior) [reply src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/VelocityUpdateModule.kt:35]: Yeah

File: src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/SpriteRenderModule.kt (hunk 3 above)
Comment (thecodewarrior) [reply src/main/java/com/teamwizardry/librarianlib/features/particlesystem/modules/SpriteRenderModule.kt:75]: I can't seem to get the binary compat checker running, but from looking at the decompiled source, as far as I can tell, it is.

**Review by Bluexin — APPROVED** [review]
